### PR TITLE
Update mx offline-first and related docs - for new offline microflow feature

### PR DESCRIPTION
**Pull request by omnajjar**
Docs update for mx8.8.0

### CODE REVIEW

**Comment by omnajjar** [comment]
Once our team is done with reviewing we can remove the draft flag, and technical writers can start their review 😃

File: content/refguide/close-page.md [hunk 1]
```diff
@@ -10,6 +10,10 @@ tags: ["studio pro"]
 This activity can be used in both **Microflows** and **Nanoflows**.
 {{% /alert %}}
 
+{{% alert type="warning" %}}
+Please note that this action will be ignored and will not have any effect when used inside a microflow that has been called from a nanoflow in an offline profile (for more information, see [Offline](offline-first#4-1-microflows)).
```
Comment (ekrem-aktas) [inline content/refguide/close-page.md]: What about this?

```suggestion
This action is ignored and does not work when microflow is called from an offline native or hybrid app. For more information, visit [Offline](offline-first#4-1-microflows)
```

File: content/refguide/committing-objects.md [hunk 2]
```diff
@@ -53,4 +53,8 @@ When committing a large number of objects, we recommend that you do not enable '
 When inside a [nanoflow](nanoflows), the object is refreshed across the client as if **Refresh in client** was set to *Yes*.
 {{% /alert %}}
 
+{{% alert type="warning" %}}
+When inside a microflow that has been called from a nanoflow in an offline profile, **Refresh in client** option will be ignored as if it was set to *No* (for more information, see [Offline](offline-first#4-1-microflows)).
```
Comment (ekrem-aktas) [inline content/refguide/committing-objects.md]: How does that look?

```suggestion
When inside a microflow that is called from an offline native or hybrid app, **Refresh in client** option is ignored and work as if it is was set to *No*. For more information, visit [Offline](offline-first#4-1-microflows)
```

File: content/refguide/microflow-call.md [hunk 3]
```diff
@@ -31,6 +31,10 @@ The microflow that is called by this activity.
 For each parameter of the microflow, you have to supply an argument of the same type. The values of the arguments are expressed using [expressions](expressions). There is a difference in the way argument values are passed to a sub-microflow:
   * Lists and objects are passed as references (meaning, if the list/object is changed in a sub-microflow, the original list/object is altered)
   * Primitive types (strings, numbers, etc.) are passed as values (meaning, they are immutable, and not changeable via sub-microflows)
+  
+{{% alert type="warning" %}}
+When used inside a nanoflow in an offline profile, only arguments of non-persistable entities that have no associations with persistable entities are allowed (for more information, see [Offline](offline-first#4-1-microflows)).
```
Comment (ekrem-aktas) [inline content/refguide/microflow-call.md]: Here it reads like we only allow NPEs and nothing else. But primitives can be passed as well.

File: content/refguide/offline-first.md [hunk 4]
```diff
@@ -132,7 +133,25 @@ Mendix helps developers in building rich offline-first apps. However, there are
...
+
+#### 4.1.1 UI Actions
+UI related actions will be ignored and will not have any effect. We encourage you to model such UI side effects in the caller nanoflow.
+
+These actions are as the following:
+* [Show message](show-message.md)
+* [Show validation message](validation-feedback.md)
+* [Show home page](show-home-page.md)
+* [Show page](show-page.md)
+* [Close page](close-page.md)
```
Comment (ekrem-aktas) [inline content/refguide/offline-first.md]: I think we also have download file action.

File: content/refguide/offline-first.md [hunk 5]
```diff
@@ -132,7 +133,25 @@ Mendix helps developers in building rich offline-first apps. However, there are
 
 ### 4.1 Microflows
 
-Microflows cannot be called directly from offline apps. However, before- and after-commit microflows still run during synchronization, which can be used for application logic on the server.
+Microflows can be called from offline apps by using [Microflow call](microflow-call.md) action in your nanoflows to perform logic on the server. However, it works a bit different that it is when used in online profiles, these differences are explained below:
+
+#### 4.1.1 UI Actions
```
Comment (ekrem-aktas) [inline content/refguide/offline-first.md]: I'd move this section after `Arguments limiations, so the order follows the microflow structure

4.1.1 Microflow Arguments
4.1.2 UI Actions
4.1.3 Microflow Return Value

Comment (ekrem-aktas) [reply content/refguide/offline-first.md]: Also, it seems we need another section to mention that when microflow works with PEs, side-effects aren't reflected.
And "Refresh in client" doesn't work even working with NPEs.

File: content/refguide/close-page.md (hunk 1 above)
Comment (omnajjar) [reply content/refguide/close-page.md]: Yes, simpler and more straight froward, I will change it! 👍

Comment (omnajjar) [reply content/refguide/close-page.md]: Did the following changes, in regard to UI Actions: 
-  Added 'Download file' action
-  Re-ordered microflow sections to put 'UI Actions' in the right order
-  Updated warning message for each action mentioned

File: content/refguide/microflow-call.md (hunk 3 above)
Comment (omnajjar) [reply content/refguide/microflow-call.md]: That's true, added ✔️

File: content/refguide/offline-first.md (hunk 4 above)
Comment (omnajjar) [reply content/refguide/offline-first.md]: Done! ✔️

File: content/refguide/offline-first.md (hunk 5 above)
Comment (omnajjar) [reply content/refguide/offline-first.md]: Added a new section for objects side-effects ✔️

Comment (omnajjar) [reply content/refguide/offline-first.md]: And for "Refresh in client" for NPEs, I think it is covered already in the new warning we added to `create-object, commit-object etc..` actions as we don't specify what type of objects but just saying this option is being used in an offline profile.

However, it is still kind of implicit, what do you think?

File: content/refguide/change-object.md [hunk 6]
```diff
@@ -65,6 +65,10 @@ When inside a [nanoflow](nanoflows), the change object action does not have the
 If **Commit type** is set to *Yes*, the object is refreshed across the client as if **Refresh in client** was set to *Yes*.
 {{% /alert %}}
 
+{{% alert type="warning" %}}
+When inside a microflow that is called from an offline native or hybrid app, **Refresh in client** option is ignored and works as if it is was set to *No*. For more information, visit [Offline](offline-first#4-1-microflows).
```
Comment (Robinfr) [inline content/refguide/change-object.md]: `as if it is was set to` -> `as if it was`

Comment (Robinfr) [inline content/refguide/change-object.md]: `visit [Offline]` -> `check out the [Offline-first](offline-first#4-1-microflows) documentation`

File: content/refguide/close-page.md [hunk 7]
```diff
@@ -10,6 +10,10 @@ tags: ["studio pro"]
 This activity can be used in both **Microflows** and **Nanoflows**.
 {{% /alert %}}
 
+{{% alert type="warning" %}}
+This action is ignored and does not work when microflow is called from an offline native or hybrid app. For more information, visit [Offline](offline-first#4-1-microflows).
```
Comment (Robinfr) [inline content/refguide/close-page.md]: `when microflow` -> `when a microflow`

File: content/refguide/change-object.md (hunk 6 above)
Comment (Robinfr) [reply content/refguide/change-object.md]: This goes for all changes.

Comment (Robinfr) [reply content/refguide/change-object.md]: Goes for all places, this was changed.

File: content/refguide/close-page.md (hunk 7 above)
Comment (Robinfr) [reply content/refguide/close-page.md]: Goes for all places.

File: content/refguide/offline-first.md [hunk 8]
```diff
@@ -132,7 +133,29 @@ Mendix helps developers in building rich offline-first apps. However, there are
 
 ### 4.1 Microflows
 
-Microflows cannot be called directly from offline apps. However, before- and after-commit microflows still run during synchronization, which can be used for application logic on the server.
+Microflows can be called from offline apps by using [Microflow call](microflow-call.md) action in your nanoflows to perform logic on the server. However, it works a bit different than it is when used in online profiles, these differences are explained below:
```
Comment (Robinfr) [inline content/refguide/offline-first.md]: `different than` -> `different from`

File: content/refguide/offline-first.md [hunk 9]
```diff
@@ -132,7 +133,29 @@ Mendix helps developers in building rich offline-first apps. However, there are
...
+These actions are as the following:
+* [Show message](show-message.md)
+* [Show validation message](validation-feedback.md)
+* [Show home page](show-home-page.md)
+* [Show page](show-page.md)
+* [Close page](close-page.md)
+* [Download file](download-file.md)
+
+#### 4.1.3 Objects Side-effects
+When your microflow is dealing with objects, changes on these objects will not be reflected to the client unless you perform synchronization in case of persistable objects, or return these objects in case thery are non-persistable objects.
```
Comment (Robinfr) [inline content/refguide/offline-first.md]: Better:

Changes to persistable objects made in a microflow will not be reflected on the client unless you synchronize the changes. Non-persistable objects have to be returned in order for changes to be reflected.

Comment (ekrem-aktas) [reply content/refguide/offline-first.md]: Do we mean sync to device with synchronize here @Robinfr ?

Comment (omnajjar) [reply content/refguide/offline-first.md]: From user perspective so far in this document, mentioning synchronize means the current synchronization, and for the seek of making stuff a little future proof and clear any ambiguity around what to sync I'll remove this part 👇 

> ...unless you synchronize ~~the changes~~ ....

File: content/refguide/close-page.md (hunk 7 above)
Comment (omnajjar) [reply content/refguide/close-page.md]: ✔️

File: content/refguide/change-object.md (hunk 6 above)
Comment (omnajjar) [reply content/refguide/change-object.md]: ✔️

Comment (omnajjar) [reply content/refguide/change-object.md]: ✔️

File: content/refguide/offline-first.md (hunk 8 above)
Comment (omnajjar) [reply content/refguide/offline-first.md]: ✔️

File: content/refguide/close-page.md (hunk 1 above)
Comment (omnajjar) [reply content/refguide/close-page.md]: ✔️

File: content/refguide/offline-first.md (hunk 9 above)
Comment (Robinfr) [reply content/refguide/offline-first.md]: Sounds good.

**Comment by omnajjar** [comment]
Okay great, I'll unmark as a draft 👍

**Comment by Adam-Dupaski** [comment]
@ConnorLand I see now this is for your component.

**Comment by ConnorLand** [comment]
> @ConnorLand I see now this is for your component.

I'll handle it. Thanks.

**Comment by ConnorLand** [comment]
I think the most recent commit on this passed build check, it's just still showing yellow on github. Therefore I think it's good to merge. 

https://travis-ci.org/github/mendix/docs/builds/668756511

**Pull request merged**